### DISCUSSION TITLE
Send HLS stream chunks to FTP Webserver (Remote FTP Storage)

### DISCUSSION THREAD

**Discussion by Vegas10128** [owncast/owncast#2861] [Ideas]
Hi Community!

So have a theory but I am not sure how to go about this. 

Owncast has S3 Storage and a few Object Storage services in the documentation. 

But all those services intend to charge after the first 1TB of data transferred. 

This would be a real deal breaker for me, 
I intend to broadcast media on 4k @ 60Hz

My issue is, After 20 simultaneous users start streaming, then the stream starts buffering. 
I know its not the server, because CPU usage is recorded around 20 to 25% during the time of streaming.
(24 Core count with 40 Gigs of Ram.) While this seems above and beyond overkill, my future needs intend to grow however this would be hurting my plans. 

This is issue is because of low upstream bandwidth. I am capped to 50Mb/s 

My goal is to accomplish the same reasons we would use S3 storage services, but instead use it for FTP protocol. 

CPanel hosting companies such as GoDaddy do not charge for metered bandwidth. 

My live stream plans are to be live 24/7 because I have loads of content 

This could be a real cost saver. Flat price no metering. 

Settings we would need --

- Endpoint
- FTP Username
- FTP Password
- Port Number
- Serving Endpoint

If anyone has ideas on how to make something like this work please let me know!

Also there are scripts such as WoWonder social network that have this exact feature for their social network, and it does work.

**Comment by LittleFox94** [comment]
Hi,

are you aware you could host your own S3-compatible storage solution, e.g. minio? At least around here (Germany) we have quite some vendors giving out unmetered servers for not-that-much money, slapping a minio on one of those would probably help you

**Reply by Vegas10128** [reply]
Hi, 
Yes I am aware, but I am also aware that I could run it as a VM, but those same companies also charge for metered bandwidth. 
That is for the United States. I cannot use Germany services due to Geo Restrictions from 2 of the networks I currently live stream at.

**Reply by LittleFox94** [reply]
alright, wanted to point out the possibility since you asked for other ideas as well :)

**Reply by Vegas10128** [reply]
yes, and thank you!

**Reply by Vegas10128** [reply]
I did a test using Minio server a few days ago as a possible solution, however it did not go very well. After setting everything up and connecting from server to minio, the stream wasnt working. Just show a still image of the stream.

**Reply by gabek** [reply]
I've certainly seen multiple people use Minio. I'm guessing there's probably a CORS or bucket configuration related issue. If you look at the errors you see in the browser console it might help point you in the right direction.

**Comment by rmens** [comment]
Why don't you use Cloudflare R2? They don't charge for traffic, only Class A/B requests. There was a large effort testing and documenting it recently.

**Reply by Vegas10128** [reply]
I have tried... multiple times and getting no results, its been a major headache to work on connecting R2 to the broadcast server. Followed the instructions here 

https://github.com/owncast/owncast.github.io/pull/112/files/c0b53b6cf763f40f6533b536701812d368017f93..2cbc0b8459a53215a956067cf1e62ef540edcb9d#diff-38fdd3747a2a9a113b93a9bb11e3b5beb8947dea352b1a4e33a84145a2e618bc

So far after setting everything up I get nothing. The player hands and presents a still image of the broadcast. 
the debug log from chrome says 
VIDEOJS: WARN: Problem encountered with playlist 0-https://website.com/hls/0/stream.m3u8. Trying again since it is the only playlist.

Logging into console and all I get is a bunch of error messages. 
ERRO[2023-03-23T15:09:17Z] Giving up on data/hls/0/stream-t84cNOB4R-30.ts
and the errors just keep on going. 
I personally never been a fan of these types of storage services because all I end up with are headaches. 

I just spent another 2 hours going through this documentation with same results, same issues. 
Ill wind up wiping my hands clean from this and find out what my options are renting a rack at a datacenter on the other side of my city. 

I did try a hacked up job using FTP with owncast under root, I had success with it between that and a web server on a local network. 
Works pretty fast, google debug console shows the responding server for the .ts files came straight from that webserver and not owncast. 

The only issue I had with this, is it involves another PC with an ftp Realtime sync program

I am trying to use CurlFtpFS but no immediate success. 
I can mount the needed directory /owncast/data/hls/0/ 

But after trying to start Owncast I get device or resource busy. 
Any suggestions on how to modify this directory and offload these files to a different folder thats shared and public please let me know!

**Reply by rmens** [reply]
Sorry to hear your experience with Cloudflare R2. I was the one validating that documentation which was written by another user. I'll take a look at any oversights.

**Reply by Vegas10128** [reply]
Thank you, its not just Cloudflare r2 I'm facing, I tried Minio as well on local network. 
I do have great networking experience but for some reason object storage on many platforms for many things have been a headache for me in general. 

Its not something I favor, because I'm more in house structure and development instead of relying on someone that can bill you every month.

**Comment by gabek** [comment]
> I can mount the needed directory /owncast/data/hls/0/
> But after trying to start Owncast I get device or resource busy.

`data/hls` is meant to be transient. Owncast recreates it every time it's launched. That's likely why you're seeing issues trying to use it this way.

**Reply by Vegas10128** [reply]
I see, thank you for letting me know about this. Explains the issue of using CurlFTPFS
